source/vir/src/ast.rs

```diff
@@ -458,6 +476,8 @@ pub enum ExprX {
     Block(Stmts, Option<Expr>),
     /// `assert_by` with a dedicated prover option (nonlinear_arith, bit_vector)
     AssertQuery { requires: Exprs, ensures: Exprs, proof: Expr, mode: AssertQueryMode },
+    /// Assertion discharged via computation
+    AssertCompute(Expr, ComputeMode),
```

Comment:
Would it make sense to use `AssertQuery` for this, and add `Compute(ComputeMode)` as a variant of `AssertQueryMode`?

Reply:
Good question.  I think it was added after I forked my branch, so I haven't looked at it closely.  I'll talk to Chanhee to learn more about whether it would work here.

Reply:
@channy412 says that he currently thinks of  `AssertQuery` as solving a specific assertion in a separate query, which is why the current signature is
```
AssertQuery { requires: Exprs, ensures: Exprs, proof: Expr, mode: AssertQueryMode }
```
In contrast, `assert_by_compute` doesn’t currently use requires or ensures, and it doesn’t spawn a separate Z3 query; it just rewrites the original assert based on the interpreter’s simplifications and then lets the normal pipleline handle the process of turning it into one or more Z3 queries.

However, we could potentially expand our notion of `AssertQuery` to signify all "special modes" of assertion handling, which would then favor a merger here.

Thoughts?

Reply:
That makes sense. Also, I had forgotten that `assert ... by` and `assert forall ... by` are encoded as `ExprX::Forall`, which I think this more closely resembles.

> it doesn’t spawn a separate Z3 query; it just rewrites the original assert based on the interpreter’s simplifications and then lets the normal pipleline handle the process of turning it into one or more Z3 queries.

Also, yes, I agree this is the important distinction.

`AssertCompute` sounds good, then. If other features with a similar structure come up, we can move to something like `AssertBy(AssertByMode)`, but it's not worth doing now.